### ISSUE TITLE
We need to deploy geonode into its own virtual server instance

### ISSUE THREAD

**Issue by timlinux** [AIFDR/inasafe#1992]
Currently we have geonode mixed in with other services on kartoza4. I prefer that it is discretely installed into its own server instance. This ticket lays out the deployment plans and requirements for Hendrik (@hvisage) to follow.

# Standard requirements

* Server installation should be scripted and repeatable with Ansible
* Virtual host installations should be scripted and repeatable with Ansible
* We should have backups in place from the outset for all components that are not ephemeral and not stored in revision control
* We should receive system health status notifications for critical events (disk full, server down etc.) and not be spammed with less critical messages causing us to ignore important messages
* Proxmox should be used for partitioning the host into virtual machines
* The procedure for setting up the machine (primarily invoking the ansible scripts) should be documented and added to the InaSAFE wiki at https://github.com/AIFDR/inasafe/wiki
* SSH password authentication should be disabled in favour of key based auth only
* General security best practices should be applied to all servers (and documented via the Ansible script)
* SSH should be run on port 8697 on all hosts.

# Provisioning

* The new server has been provisioned with 32GB RAM and 2 x 3TB enterprise disks
* We need to request additional IP addresses for the VM's unless we can use masquerading with proxmox

# Architecture

The system should be divided into 3 virtual machines:

* Linux server running  Ubuntu 14.04 running docker name: **docker.inasafe.org**
* Linux server running Ubuntu 14.04 name: **geonode.inasafe.org**
* Windows server running Windows 64 bit installed with Jenkins - this is lower priority but we should reserve space for the vm in our planning

## ProxMox platform

Proxmox GUI: http://proxmox1.inasafe.org:8006

## Docker VM

* The docker VM should be installed with a bare bones 14.04 installation
* This VM should be allocated 1TB of disk space
* This VM should be allocated 16 GB of RAM
* Docker should be installed from the upstream docker repos (not the out of date packages provided by ubuntu)
* docker-compose must be installed
* nginx should be installed via apt and the Nginx configurations form kartoza4 should be copied over and deployed
* docker based services should be spun up
* DNS entries currently pointing to kartoza4 should be updated to point to this VM
* Tim to provide input on deployment of the additional Docker containers

### DNS

Record | Notes
----------|----------------
inasafe.org | Main website for InaSAFE running wordpress in docker
staging.inasafe.org | Test website for InaSAFE running wordpress in docker
docs.inasafe.org | Main documentation repository for training materials (master)
staging.docs.inasafe.org | Staging documentation repository for training materials (develop)
changelog.inasafe.org | Production site for compiling changelogs for each release
data.inasafe.org | Download service for sample datasets. Datasets provisioned with btsync. To be replaced with GeoNode.
experimental.inasafe.org | Experimental plugin repository for InaSAFE QGIS plugin
floods.inasafe.org | Production site for flood reporting system written in django, deployed in docker.
staging.floods.inasafe.org | Staging site for flood reporting system written in django, deployed in docker.
osm.inasafe.org | Production download server for OSM data used in InaSAFE, written in flask
users.inasafe.org | Production user map web site. Written in django, deployed in docker.
staging.inasafe.org | Staging user map web site. Written in django, deployed in docker.
realtime.inasafe.org | Production realtime earthquake impact report. This is a **critical** service. Currently static html, will soon be django in docker.

### Ports

Port | Service | Purpose
------|-------------|-------------
8697 | SSH     | Sysadmin access to host using keys only
80 | Web | Public access to the inasafe websites. Traffic on this port should redirect to 443. Nginx will reverse proxy traffic into the host.
443 | TLS/Web | Public access to the inasafe websites. Nginx will reverse proxy traffic into the host.
9222 | SFTP | Used to publish shakemaps in near realtime from BMKG in Indonesia to our server. This is  **critical** service.
???? | btsync | Used to synchronise datasets for realtime. In their docker container, btsync services use 55555 and 8888, but these are not exposed to the host. However this protocol's traffic needs to be allowed in & out of the server.

### Mounts
PostgreSQL backup target: /data/dbbackups 
Geoserver data dir: /usr/share/geoserver/data

## Geonode VM (ref #1241)

* The geonode VM should be installed with a bare bones 14.04 installation
* This VM should be allocated 1TB of disk space
* This VM should be allocated 8 GB of RAM
* Geonode should be deployed from packages to this server
* @gubuntu to provide guidance as to the scripting of the installation procedure

### DNS

Record | Notes
----------|----------------
geonode.inasafe.org | Data portal for InaSAFE downloadable data
geoserver.inasafe.org | Map server backend for InaSAFE downloadable data

Comment from Gavin: The geoserver sub-domain is unnecessary since geoserver will be at geonode.inasafe.org/geoserver

### Ports

Port | Service | Purpose
------|-------------|-------------
8697 | SSH     | Sysadmin access to host using keys only
80 | Web | Public access to the geonode website. Traffic on this port should redirect to 443
443 | TLS/Web | Public access to the geonode website.
8080 | Web | Public access to the geoserver website. Needed for Geonode to function. This is reverse-proxied from 8080 internally (i.e. it's not necessary to configure)

## Geonode Testing VM

* The testing geonode VM should be installed with a bare bones 14.04 installation
* This VM should be allocated 500GB of disk space
* This VM should be allocated 4 GB of RAM
* Geonode should be deployed from packages to this server
* @gubuntu to provide guidance as to the scripting of the installation procedure
* DNS entry for staging.geonode.insasafe.org should be pointed to this server

### DNS

Record | Notes
----------|----------------
staging.geonode.inasafe.org | Data portal for InaSAFE downloadable data
staging.geoserver.inasafe.org | Map server backend for InaSAFE downloadable data

Comment from Gavin: The geoserver sub-domain is unnecessary since geoserver will be at staging.geonode.inasafe.org/geoserver

### Ports

Port | Service | Purpose
------|-------------|-------------
8697 | SSH     | Sysadmin access to host using keys only
80 | Web | Public access to the geonode test website. Traffic on this port should redirect to 443
443 | TLS/Web | Public access to the test geonode website.
8080 | Web | Public access to the geoserver website. Needed for Geonode to function. This is reverse-proxied from 8080 internally (i.e. it's not necessary to configure)

## Windows VM

* The VM should be allocated 4GB of RAM
* This VM should be allocated 500GB of disk space
* The VM should be installed with a bare bones Windows 8 installation
* We will install Jenkins in the windows VM and set it autorun tests
* This VM is lowest priority

### DNS

Record | Notes
----------|----------------
windows.jenkins.inasafe.org | Windows jenkins server for InaSAFE

### Ports
Port | Service | Purpose
------|-------------|-------------
????  | RDP     | Sysadmin access to host over secure tunnel or VPN. @hvisage to advise on options
443 | TLS/Web | Public access to the Jenkins website. Still to be determined if we need this but I think we will.

# Public ssh keys

* All VM's set up should include Tim's ssh public keys and an account with username 'timlinux' set up for me. Tim will send ssh key separately. 
* And user 'gavin' on both geonode VMs.
* And these users on docker and geonode VMs:
** Richard
** Ismail 
** Rizky
** Christian
** Jannes 
** Etienne
** Maria

## Firewall

firewall GUI: http://vfw1.inasafe.org:4433 (https://5.9.166.170:4433/ until DNS is configured)

### IP addresses and NAT config

VM | LAN | LAN ssh port | WAN
------|-------------|------------|---------------
docker | 10.10.10.10 | 2022 | 5.9.160.107
geonode | 10.10.10.11 | 2122 | 5.9.160.105
geonode-stage | 10.10.10.12 | 2222 | 5.9.160.106

# Network Diagram

https://www.lucidchart.com/invitations/accept/a3e17101-934b-4ec3-88dd-a50c3f524ef6

![inasafe network diagram - new page](https://cloud.githubusercontent.com/assets/178003/8062093/224b4442-0ef8-11e5-9465-fe4eb378c5c7.png)

# CC

@hvisage
@gubuntu

**Comment by hvisage** [comment]
Public accessible Ports to be setup?

Windows: RDP-3389
Geonode?
Docker: SSH-8967
 Which others?

The Idea will be a single firewall that will NAT accordingly

**Comment by gubuntu** [comment]
@timlinux where should the ansible code go, Somewhere in the InaSAFE repo?

**Comment by timlinux** [comment]
@gubuntu @hvisage @rduivenvoorde

We have this repo that we should use for maintaining configuration and deployment scripts.

https://github.com/AIFDR/inasafe-infrastructure

**Comment by timlinux** [comment]
> Public accessible Ports to be setup?

@hvisage I have update the original ticket description with more details on this - we will obviously need to be able to adjust the rules over time as we add / reconfigure services.

Also note that I am changing the spec above to add a fourth partition that we can use for a geonode sandbox / testing server.

**Comment by timlinux** [comment]
**Notes:** 

* Before closing this ticket, the details above should be transposed to the wiki
* We still need to make a plan for Jenkins (running on Linux)

**Comment by gubuntu** [comment]
@timlinux, a migration plan and time window needs to be scheduled for this. Since proxmox runs on a modified debian kernel, the host machine and VMs will need to be set up from scratch. Can/must the existing services be migrated to another host (kartoza3?) during setup, or can downtime be scheduled and services restored from backups / docker images / ansible?

**Comment by gubuntu** [comment]
@timlinux, you refer to a new server, i.e. not kartoza4? Please provide IP etc. of the new server. Then, what is the fate of kartoza4?

**Comment by timlinux** [comment]
@gubuntu No usable IP yet until @hvisage is done with setting up proxmox and the basic installations, firewall etc on it.

Kartoza4 will be used for jenkins / testing and I will decommission kartoza2

**Comment by hvisage** [comment]
firewall GUI: http://vfw1.inasafe.org:4433
Proxmox GUI: http://proxmox1.inasafe.org:8006

**Comment by hvisage** [comment]
Creating a 20GB base Image using Ubuntu 14.04 LTS that'll be cloned for the rest

**Comment by hvisage** [comment]
Users (with keys from t4a) created on proxmox1.inasafe.org:
hvisage@proxmox1:~$ ls /home
admire  gavin  hvisage  jurgen  robert  tom

Could somebody please upload the required Windows .ISO and let me know in which directory they dumped it?

**Comment by timlinux** [comment]
Hi

Please remove admire, jurgen, robert, tom - note that this project is completely unrelated to T4A so no overlap in users except for myself and gavin. We will need to add accounts for 

* Richard
* Ismail 
* Rizky
* Christian
* Jannes 
* Etienne
* Maria

 but we can sort the others out later.

> Could somebody please upload the required Windows .ISO and let me know in which directory they dumped it?

I will upload it next week when I get back to SA.

Thanks!

**Comment by hvisage** [comment]
Created pull request for my current ansible for the proxmox settings

**Comment by hvisage** [comment]
three VMs created, and NATted via vfw1.inasafe.org (5.9.166.170)
docker 10.10.10.10 - sshport: 2022
geonode 10.10.10.11 - sshport: 2122
geonode-stage 10.10.10.12 - sshport: 2222

only the user "ubuntu" configured. If you have ansible playbooks ready for configuring them, please do/share.

so that I can incorporate the storage and mount them in the right places

**Comment by gubuntu** [comment]
@hvisage I've added geonode guidelines to https://github.com/AIFDR/inasafe-infrastructure for you to work with. 

I can't reach the firewall and proxmox GUIs you refer to above...?

**Comment by hvisage** [comment]
https://5.9.166.170:4433/ -> Tim still haven't added the DNS for
vfw1.inasafe.org
https://proxmox1.inasafe.org:8006/ <- This one definitely works.

On Wed, Jun 10, 2015 at 6:58 PM, Gavin Fleming <notifications@github.com>
wrote:

> @hvisage <https://github.com/hvisage> I've added geonode guidelines to
> https://github.com/AIFDR/inasafe-infrastructure for you to work with.
>
> I can't reach the firewall and proxmox GUIs you refer to above...?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/AIFDR/inasafe/issues/1992#issuecomment-110834369>.
>



-- 
---
Hendrik Visage
Envisage Systems / Envisage Cloud Solutions
084-612-5345 or 083-675-5506
hvisage@envisage.co.za

**Comment by hvisage** [comment]
PS: check the pull request for my ansible prelimary stuff

On Wed, Jun 10, 2015 at 7:04 PM, H Visage <hvisage@envisage.co.za> wrote:

> https://5.9.166.170:4433/ -> Tim still haven't added the DNS for
> vfw1.inasafe.org
> https://proxmox1.inasafe.org:8006/ <- This one definitely works.
>
> On Wed, Jun 10, 2015 at 6:58 PM, Gavin Fleming <notifications@github.com>
> wrote:
>
>> @hvisage <https://github.com/hvisage> I've added geonode guidelines to
>> https://github.com/AIFDR/inasafe-infrastructure for you to work with.
>>
>> I can't reach the firewall and proxmox GUIs you refer to above...?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/AIFDR/inasafe/issues/1992#issuecomment-110834369>.
>>
>
>
>
> --
> ---
> Hendrik Visage
> Envisage Systems / Envisage Cloud Solutions
> 084-612-5345 or 083-675-5506
> hvisage@envisage.co.za
>



-- 
---
Hendrik Visage
Envisage Systems / Envisage Cloud Solutions
084-612-5345 or 083-675-5506
hvisage@envisage.co.za

**Comment by hvisage** [comment]
Added NAT rules:
[image: icon]
<https://5.9.166.170:4433/firewall_nat_1to1.php?act=toggle&id=0> WAN
5.9.160.105 10.10.10.11 * geonode     [image: edit]
<https://5.9.166.170:4433/firewall_nat_1to1_edit.php?id=0>   [image: delete]
<https://5.9.166.170:4433/firewall_nat_1to1.php?act=del&id=0>  [image:
duplicate] <https://5.9.166.170:4433/firewall_nat_1to1_edit.php?dup=0>
  [image:
icon] <https://5.9.166.170:4433/firewall_nat_1to1.php?act=toggle&id=1> WAN
5.9.160.106 10.10.10.12 * geonode-stage     [image: edit]
<https://5.9.166.170:4433/firewall_nat_1to1_edit.php?id=1>   [image: delete]
<https://5.9.166.170:4433/firewall_nat_1to1.php?act=del&id=1>  [image:
duplicate] <https://5.9.166.170:4433/firewall_nat_1to1_edit.php?dup=1>
  [image:
icon] <https://5.9.166.170:4433/firewall_nat_1to1.php?act=toggle&id=2> WAN
5.9.160.107 10.10.10.10 * docker     [image: edit]
<https://5.9.166.170:4433/firewall_nat_1to1_edit.php?id=2>



Added rules for ports 22, 8679, 80 and 443 to those servers.

Please advise if there are anything more needed at present to continue
working?

I also haven't seen the exact directories I need to mount the storage on,
so please point me that way please ;)




On Wed, Jun 10, 2015 at 7:05 PM, H Visage <hvisage@envisage.co.za> wrote:

> PS: check the pull request for my ansible prelimary stuff
>
> On Wed, Jun 10, 2015 at 7:04 PM, H Visage <hvisage@envisage.co.za> wrote:
>
>> https://5.9.166.170:4433/ -> Tim still haven't added the DNS for
>> vfw1.inasafe.org
>> https://proxmox1.inasafe.org:8006/ <- This one definitely works.
>>
>> On Wed, Jun 10, 2015 at 6:58 PM, Gavin Fleming <notifications@github.com>
>> wrote:
>>
>>> @hvisage <https://github.com/hvisage> I've added geonode guidelines to
>>> https://github.com/AIFDR/inasafe-infrastructure for you to work with.
>>>
>>> I can't reach the firewall and proxmox GUIs you refer to above...?
>>>
>>> —
>>> Reply to this email directly or view it on GitHub
>>> <https://github.com/AIFDR/inasafe/issues/1992#issuecomment-110834369>.
>>>
>>
>>
>>
>> --
>> ---
>> Hendrik Visage
>> Envisage Systems / Envisage Cloud Solutions
>> 084-612-5345 or 083-675-5506
>> hvisage@envisage.co.za
>>
>
>
>
> --
> ---
> Hendrik Visage
> Envisage Systems / Envisage Cloud Solutions
> 084-612-5345 or 083-675-5506
> hvisage@envisage.co.za
>



-- 
---
Hendrik Visage
Envisage Systems / Envisage Cloud Solutions
084-612-5345 or 083-675-5506
hvisage@envisage.co.za

**Comment by timlinux** [comment]
Please remove 22 - we run ssh on 8697 as standard practice

@gubuntu please advice which directories he should mount for storage I guess the geoserver data dir would be the main candidate...

**Comment by hvisage** [comment]
Once the ansible scripts have been run then yes. And running on a different
port is just a obfuscation, that us why I install and run sshguard

---
replied from my small phone :)
--
Hendrik Visage

**Comment by gubuntu** [comment]
I normally put geoserver data in /data/geoserver_data. And a target for PostgreSQL dumps, like /data/dbbackups. 

Ansible can use any ssh port specified (see how in T4A repo), that doesn't have to be a blocker.

**Comment by timlinux** [comment]
@gubuntu I was referring specifically to where geonode dumps it

**Comment by hvisage** [comment]
@gubuntu I've created the VMs, and I'm going to take a step from ansible as
my attempts to hammer T4A's into InaSafe submission have failed so I first
need to familiarize myself with what/where/etc. That is why I've tried to
say: you can go ahead with your ansible stuff to deploy if you know how and
have that fix the SSH port(s) etc. to your likings, I've set things up for
you to start working and not be waiting on me to finish learning with
Ansible... just one of those things: when I'm idle, nobody needs me, when
I'm busy everybody needs my time ;(

will try to get away from here by lunch time to setup the geonode's storage
to /data
@tim you've mentioned /var/lib/docker for docker??

On Thu, Jun 11, 2015 at 10:57 AM, Gavin Fleming <notifications@github.com>
wrote:

> I normally put geoserver data in /data/geoserver_data. And a target for
> PostgreSQL dumps, like /data/dbbackups.
>
> Ansible can use any ssh port specified (see how in T4A repo), that doesn't
> have to be a blocker.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/AIFDR/inasafe/issues/1992#issuecomment-111049079>.
>



-- 
---
Hendrik Visage
Envisage Systems / Envisage Cloud Solutions
084-612-5345 or 083-675-5506
hvisage@envisage.co.za

**Comment by timlinux** [comment]
> three VMs created, and NATted via vfw1.inasafe.org (5.9.166.170)
> docker 10.10.10.10 - sshport: 2022
> geonode 10.10.10.11 - sshport: 2122
> geonode-stage 10.10.10.12 - sshport: 2222

Please see the notes above - I asked for ssh to run on port 8697 and pub/private key authentication to be enabled only on all hosts.

> @tim you've mentioned /var/lib/docker for docker?

Yes it should be a large volume say 400gb and the remaining space for the /home

Can you please summarise everything nicely - I can't pick through the comments trail easily to find the needed info. Rather get everything set up and then give all the info in one go. And please work with @rduivenvoorde to get him logged in and needed privileged's on the docker host, and fill him in on where the playbook stuff is.

**Comment by hvisage** [comment]
On Thu, Jun 11, 2015 at 11:47 AM, Tim Sutton <notifications@github.com>
wrote:

> three VMs created, and NATted via vfw1.inasafe.org (5.9.166.170)
> docker 10.10.10.10 - sshport: 2022
> geonode 10.10.10.11 - sshport: 2122
> geonode-stage 10.10.10.12 - sshport: 2222
>
> Please see the notes above - I asked for ssh to run on port 8697 and
> pub/private key authentication to be enabled only on all hosts.
>

Tim, I've been unexpectedly busier the past 2-3weeks, than the previous
6months. Yes I know, but I've setup those access to get to VMs for
Gavin/Rduivenvoorde/etc. to be able to log and prep while we were awaiting
the extra IPs, and the Ansible playbook to do that setup.

>  @tim <https://github.com/tim> you've mentioned /var/lib/docker for
> docker?
>
> Yes it should be a large volume say 400gb and the remaining space for the
> /home
>
> Can you please summarise everything nicely - I can't pick through the
> comments trail easily to find the needed info. Rather get everything set up
> and then give all the info in one go. And please work with @rduivenvoorde
> <https://github.com/rduivenvoorde> to get him logged in and needed
> privileged's on the docker host, and fill him in on where the playbook
> stuff is.
>

I'm replying via email as I have the time to have it at least documented.

The only playbook stuff I have, is some attempts to configure the Proxmox1,
and even that is not doing the ssh port 8697 as I expected it to do. I'll
look at that later this weekend/next week with the public holiday to
breath.

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/AIFDR/inasafe/issues/1992#issuecomment-111069876>.
>



-- 
---
Hendrik Visage
Envisage Systems / Envisage Cloud Solutions
084-612-5345 or 083-675-5506
hvisage@envisage.co.za

**Comment by timlinux** [comment]
Its all up and running on http://geonode.inasafe.org

**Comment by abersh** [comment]
I want to deploy geonode for production and I wnat to deploy geoserver in another server(machine) and also postgis in another server (machine) is it possible? how?

**Comment by hvisage** [comment]
It should be. The system as deployed as separate dockers, to help keep the functional parts separate.
 (but rather open a different thread for this type of discussion ;) )

**Comment by abersh** [comment]
@hvisage thank you !!!
